Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp

```diff
@@ -0,0 +1,475 @@
+                    continue;
+                widget->remove_from_parent();
+                widget->set_visible(true);
+                (void)reciever->try_add_child(*widget);
+            }
+        };
+
+        transfer_childs(root_container, child_widgets());
+
+        m_detached_widgets_window->on_close = [this, root_container, transfer_childs]() {
+            transfer_childs(this, root_container->child_widgets());
+            this->set_view_state(ViewState::Expanded);
+            unregister_open_window(m_detached_widgets_window);
+            this->m_detached_widgets_window = nullptr;
+        };
+    }
+
+    register_open_window(m_detached_widgets_window);
+
+    if (m_is_dragging)
+        m_detached_widgets_window->move_to(screen_relative_rect().location().translated(m_current_mouse_position).translated({ m_detached_widgets_window->width() / -2, 0 }));
+    m_detached_widgets_window->show();
+    return {};
+}
+
+void DynamicWidgetContainer::close_all_detached_windows()
+{
+    for (auto window : DynamicWidgetContainer::s_open_windows.in_reverse()) {
+        window->close();
+    }
+}
+
+void DynamicWidgetContainer::register_open_window(RefPtr<GUI::Window> window)
+{
+    DynamicWidgetContainer::s_open_windows.append(window);
+}
+
+void DynamicWidgetContainer::unregister_open_window(RefPtr<GUI::Window> window)
+{
+    Optional<size_t> match = DynamicWidgetContainer::s_open_windows.find_first_index(window);
+    if (match.has_value())
+        DynamicWidgetContainer::s_open_windows.remove(match.value());
+}
+
+void DynamicWidgetContainer::mousedown_event(MouseEvent& event)
+{
+    if (event.button() != MouseButton::Primary)
+        return;
+
+    if (m_expand_collapse_button.contains(event.position()) && !is_container_with_individual_order())
+        set_view_state(view_state() == ViewState::Expanded ? ViewState::Collapsed : ViewState::Expanded);
+
+    if (m_detach_widget_button.contains(event.position()))
+        set_view_state(ViewState::Detached);
+}
+
+void DynamicWidgetContainer::mouseup_event(MouseEvent& event)
+{
+    if (event.button() != MouseButton::Primary)
+        return;
+
+    if (m_is_dragging) {
+        // If we dropped the widget outside of ourself, we would like to detach it.
+        if (!m_parent_container.has_value() && !rect().contains(event.position()))
+            set_view_state(ViewState::Detached);
+
+        if (m_parent_container.has_value()) {
+            bool should_move_position = m_parent_container.value()->check_has_move_target(relative_position().translated(m_current_mouse_position), MoveTargetOp::ClearAllTargets);
+
+            if (should_move_position)
+                m_parent_container.value()->swap_widget_positions(this, relative_position().translated(m_current_mouse_position));
+            else
+                set_view_state(ViewState::Detached);
+        }
+
+        m_is_dragging = false;
+        set_override_cursor(Gfx::StandardCursor::Arrow);
```

Comment:
This seems unnecessary and wrong.

Reply:
If that is not in place, the mouse cursor will only change back if you move the mouse cursor.